### PR TITLE
First pass of multipart file uploading

### PR DESCRIPTION
**Pull request by mikeswoods**

### CODE REVIEW

File: src/bf/api/client/mod.rs [hunk 1]
```diff
@@ -335,98 +323,35 @@ impl Blackfynn {
...
     }
 
     /// Generate a preview of the files to be uploaded.
-    pub fn preview_upload<P, Q>(&self, path: P, files: &Vec<Q>, append: bool) -> bf::Future<response::PreviewPackage>
+    pub fn preview_upload<P, Q>(&self, path: P, files: &[Q], append: bool) -> bf::Future<response::PreviewPackage>
         where P: AsRef<Path>,
               Q: AsRef<Path>
     {
         // Parition the S3 files into two groups, successes and failures:
-        let (good, bad): (Vec<_>, Vec<_>) = files.into_iter().map(|file| model::S3File::new(path.as_ref(), file.as_ref())).partition(Result::is_ok);
```
Comment (scirner22) [inline src/bf/api/client/mod.rs]: looks like the comment above is no longer valid

File: src/bf/api/client/mod.rs [hunk 2]
```diff
@@ -335,98 +323,35 @@ impl Blackfynn {
...
-        // If a filaure occurred, return immediately:
-        let mut errs: Vec<bf::error::Error> = bad.into_iter().map(Result::unwrap_err).collect();
-        if let Some(err) = errs.pop() {
-            return Box::new(future::err(err));
-        }
-
-        let s3_files: Vec<model::S3File> = good.into_iter().map(Result::unwrap).collect();
+        let s3_files = match results {
+            Ok(good) => good,
+            Err(e) => return into_future_trait(future::err(e))
```
Comment (scirner22) [inline src/bf/api/client/mod.rs]: didn't know you can return like this!

File: src/bf/api/client/mod.rs [hunk 3]
```diff
@@ -335,98 +323,35 @@ impl Blackfynn {
...
-                })
-        }))
-        .into_future()
-        .map_err(|(e, _)| e)
-        .and_then(|_| Ok(import_id));
-
-        Box::new(f)
+    /// Returns a S3 uploader.
+    pub fn s3_uploader(&self, creds: &TemporaryCredential) -> S3Uploader {
+        S3Uploader::new(self.inner.borrow().config.s3_server_side_encryption().clone(),
```
Comment (scirner22) [inline src/bf/api/client/mod.rs]: do you mind searching this whole namespace for self.inner.borrows and borrow_muts and verify none are moved into futures? That's the only case I can think of where we could get a weird panic case.

Comment (mikeswoods) [reply src/bf/api/client/mod.rs]: Good call. I think everything looks good

**Pull request merged**